File: examples/projects/ruby/lib/calculator.rb
class Calculator
  def add(a, b)
    a + b
  end

  def subtract(a, b)
    a - b
  end

  def multiply(a, b)
    a * b
  end

  def divide(a, b)
    raise ArgumentError, 'Cannot divide by zero' if b.zero?
    a.to_f / b
  end

  def power(base, exponent)
    base**exponent
  end

  def square_root(n)
    raise ArgumentError, 'Cannot calculate square root of negative number' if n < 0
    Math.sqrt(n)
  end

  def factorial(n)
    raise ArgumentError, 'Factorial is not defined for negative numbers' if n < 0
    return 1 if n <= 1
    (1..n).reduce(:*)
  end

  def average(numbers)
    return 0 if numbers.empty?
    numbers.sum.to_f / numbers.size
  end
end

File: examples/projects/javascript/src/calculator.js
class Calculator {
  add(a, b) {
    return a + b;
  }

  subtract(a, b) {
    return a - b;
  }

  multiply(a, b) {
    return a * b;
  }

  divide(a, b) {
    if (b === 0) {
      throw new Error('Division by zero is not allowed');
    }
    return a / b;
  }

  power(base, exponent) {
    return Math.pow(base, exponent);
  }

  sqrt(n) {
    if (n < 0) {
      throw new Error('Cannot calculate square root of negative number');
    }
    return Math.sqrt(n);
  }

  factorial(n) {
    if (n < 0) {
      throw new Error('Factorial is not defined for negative numbers');
    }
    if (n === 0 || n === 1) {
      return 1;
    }
    let result = 1;
    for (let i = 2; i <= n; i++) {
      result *= i;
    }
    return result;
  }

  average(numbers) {
    if (!Array.isArray(numbers) || numbers.length === 0) {
      throw new Error('Invalid input: array of numbers required');
    }
    const sum = numbers.reduce((acc, num) => acc + num, 0);
    return sum / numbers.length;
  }
}

module.exports = Calculator;